Adds DNSControllerSpec and WatchIngress flag

geojaz opened the pull request:
This PR is in reference to #2496, #2468 and the issues referenced in there relating to use of the watch-ingress flag. 

This PR attempts to rectify this situation and gives users who want it, the option to turn on watch-ingress without forcing it on them. Also spits out a warning to the logs about potential side effects.

Includes notes in `docs/cluster_spec.md` to explain.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kops/2504)
<!-- Reviewable:end -->

geojaz commented:
@justinsb Minor API changes for a good cause here :) Would you pleas take a look?

justinsb commented:
I like this, but can we just revert the change so we can get 1.6.0 out?

justinsb commented:
I'm also wondering if we adopt external-dns for ingress

geojaz commented:
Yes please, let's get 1.6 out

DerekV commented:
Any plans to include this in the next release?

chrislovecnm commented:
@justinsb we good to get this in?

chrislovecnm commented:
@geojaz need a rebase

chrislovecnm commented:
@geojaz status on this?

geojaz commented:
@chrislovecnm  I'll try to get this done tonight

geojaz commented:
@chrislovecnm :)

justinsb commented on upup/pkg/fi/cloudup/template_functions.go:
```diff
@@ -128,11 +130,21 @@ func (tf *TemplateFunctions) GetInstanceGroup(name string) (*kops.InstanceGroup,
 
 	argv = append(argv, "/usr/bin/dns-controller")
 
+	if tf.cluster.Spec.DNSController != nil {
```
FYI, I tend to do this:

```
dnsController := tf.cluster.Spec.DNSController
if dnsController == nil {
  dnsController = &DNSControllerSpec{}
}
...
```

With one option it's a wash, with a bunch it makes things simpler :-)

justinsb commented:
This LGTM.  I'm not sure if we should call it `dnsController` or `externalDns` though...  I'll leave that up to you @geojaz 

My thoughts:

* eventually kops will use external-dns
* external-dns makes it clearer that this is not kube-internal DNS

But:

* it doesn't really matter anyway; we can change in future and both are reasonably clear :-)

justinsb commented:
/lgtm

geojaz commented:
I think we should call it externalDns with some language on the docs/commenting in the code to explain that this is also the config for dnsController. I'll clean this up and let you take one more look at it before merge. Thanks for the ideas :)

kenden commented on docs/cluster_spec.md:
````diff
@@ -131,6 +131,17 @@ spec:
+  watchIngress: true
+```
+
+Default kops behavior is false. `watchIngress: true` uses the default _dns-controller_ behavior which is to watch the ingress controller for changes. Set this option at risk of interrupting Service updates in some cases.
````
"Set this option at risk of interrupting Service updates in some cases.".
It would be nice to add more info about when this would/could happen.

geojaz commented:
Will squash commits as make sense. I moved the config from DNSControllerSpec to ExternalDNSConfig and updated the docs and such. I think next step is to refactor to combine the DNSControllerArgv and ExternalDNSArgv in template_functions... but maybe not crucial for this submission.

chrislovecnm commented:
Need a rebase and e2e failed.

chrislovecnm commented:
/lgtm cancel

chrislovecnm commented:
/lgtm

DerekV commented:
Is there some way to move this forward if it just needs a rebase?

chrislovecnm commented:
Paging @geojaz :)

geojaz commented:
/lgtm
can i lgtm my own pr? ;)

chrislovecnm commented:
/lgtm

@DerekV will be in our next release, let us know if you have any questions about using it before the release
